runtime/src/tiered_storage/writer.rs

```diff
@@ -0,0 +1,86 @@
+//! docs/src/proposals/append-vec-storage.md
+
+use {
+    crate::{
+        account_storage::meta::{StorableAccountsWithHashesAndWriteVersions, StoredAccountInfo},
+        storable_accounts::StorableAccounts,
+        tiered_storage::{
+            error::TieredStorageError,
+            file::TieredStorageFile,
+            footer::{AccountMetaFormat, TieredStorageFooter},
+            hot::HotAccountMeta,
+            meta::TieredAccountMeta,
+            TieredStorageFormat, TieredStorageResult,
+        },
+    },
+    solana_sdk::{account::ReadableAccount, hash::Hash},
+    std::{borrow::Borrow, path::Path},
+};
+
+#[derive(Debug)]
+pub struct TieredStorageWriter<'format> {
+    storage: TieredStorageFile,
+    format: &'format TieredStorageFormat,
+}
+
+impl<'format> TieredStorageWriter<'format> {
+    pub fn new(
+        file_path: impl AsRef<Path>,
+        format: &'format TieredStorageFormat,
+    ) -> TieredStorageResult<Self> {
+        Ok(Self {
+            storage: TieredStorageFile::new_writable(file_path)?,
+            format,
+        })
+    }
+
+    #[inline]
+    fn write_accounts_impl<
+        'a,
+        'b,
+        T: ReadableAccount + Sync,
+        U: StorableAccounts<'a, T>,
+        V: Borrow<Hash>,
+        W: TieredAccountMeta,
+    >(
+        &self,
+        _accounts: &StorableAccountsWithHashesAndWriteVersions<'a, 'b, T, U, V>,
+        footer: TieredStorageFooter,
+        mut _account_metas: Vec<W>,
+        _skip: usize,
+    ) -> TieredStorageResult<Vec<StoredAccountInfo>> {
+        footer.write_footer_block(&self.storage)?;
+
+        Err(TieredStorageError::Unsupported())
+    }
+
+    pub fn write_accounts<
+        'a,
+        'b,
+        T: ReadableAccount + Sync,
+        U: StorableAccounts<'a, T>,
+        V: Borrow<Hash>,
+    >(
+        &self,
+        accounts: &StorableAccountsWithHashesAndWriteVersions<'a, 'b, T, U, V>,
+        skip: usize,
+    ) -> TieredStorageResult<Vec<StoredAccountInfo>> {
+        let footer = TieredStorageFooter {
+            account_meta_format: self.format.account_meta_format,
+            owners_block_format: self.format.owners_block_format,
+            account_block_format: self.format.account_block_format,
+            account_index_format: self.format.account_index_format,
+            ..TieredStorageFooter::default()
+        };
+        match footer.account_meta_format {
+            AccountMetaFormat::Hot => {
+                self.write_accounts_impl(accounts, footer, Vec::<HotAccountMeta>::new(), skip)
```

Comment:
Oh oh, the main point is that different formats will invoke the same function but with different TieredStorageMeta formats.  (For instance, Hot storage will use HotAccountMeta, while Cold storage will use ColdAccountMeta).

If we want to remove one level, then the match will be moved to TieredStorage::write_accounts.  But I feel hiding the detail under Writer is cleaner since this is the implementation detail under the Writer.